File: Leetcode/94_2.cpp
/**
 * 94. 二叉树的中序遍历 Morris Traversal
 * link：https://leetcode-cn.com/problems/binary-tree-inorder-traversal/
 */

class Solution
{
public:
    vector<int> inorderTraversal(TreeNode *root)
    {
        vector<int> result;
        TreeNode *predecessor;

        while (root != nullptr)
        {
            if (root->left == nullptr)
            {
                result.push_back(root->val);
                root = root->right;
            }
            else
            {
                predecessor = root->left;
                while (predecessor->right != nullptr && predecessor->right != root)
                    predecessor = predecessor->right;

                if (predecessor->right == nullptr)
                {
                    predecessor->right = root;
                    root = root->left;
                }
                else
                {
                    result.push_back(root->val);
                    predecessor->right = nullptr;
                    root = root->right;
                }
            }
        }

        return result;
    }
};

File: Leetcode/94_2.py
# 94. 二叉树的中序遍历  Morris Traversal
# link：https://leetcode-cn.com/problems/binary-tree-inorder-traversal/

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        result = []

        while root:
            if root.left:
                predecessor: TreeNode = root.left
                while predecessor.right and predecessor.right != root:
                    predecessor = predecessor.right

                if not predecessor.right:
                    predecessor.right = root
                    root = root.left
                else:
                    result.append(root.val)
                    predecessor.right = None
                    root = root.right
            else:
                result.append(root.val)
                root = root.right

        return result
